### PR TITLE
__Property__ Parsing + Decorated Sink Context + Ordered Sink Comparator

### PR DESCRIPTION
**Pull request by TanayParikh**
- Updated Java JSON parsing to use `ExprParser` instead of `FObjectParser`. The `ExprParser` has the `FObjectParser` as one of the `Alt` options.
- Added functionality for decorated sinks to have context
- Added explicit comparator to ordered sink `eof` sorting
- Java JSON array parsing (null class) bug fix

### CODE REVIEW

File: src/foam/dao/AbstractDAO.java [hunk 1]
```diff
@@ -61,6 +61,13 @@ protected Sink decorateSink_(Sink sink, long skip, long limit, Comparator order,
     return sink;
   }
 
+  protected Sink decorateSink_(X x, Sink sink, long skip, long limit, Comparator order, Predicate predicate) {
```
Comment (kgrgreer) [inline src/foam/dao/AbstractDAO.java]: Just move the body of the other decorateSink_ (the one without the X) here and remove the other.

File: src/foam/dao/SinkJava.js [hunk 2]
```diff
@@ -111,21 +111,23 @@ foam.CLASS({
...
     {
       name: 'put',
-      javaCode: 'if ( getCount() >= getLimit() ) {\n'
-              + '  if ( sub != null ) sub.detach();\n'
-              + '} else {\n'
-              + '  setCount(getCount() + 1);\n'
-              + '  getDelegate().put(obj, sub);\n'
-              + '}\n'
+      javaCode: 
+`if ( getCount() >= getLimit() ) {
```
Comment (kgrgreer) [inline src/foam/dao/SinkJava.js]: I prefer this indented over.

File: src/foam/dao/SinkJava.js [hunk 3]
```diff
@@ -111,21 +111,23 @@ foam.CLASS({
...
     },
     {
       name: 'remove',
-      javaCode: 'if ( getCount() >= getLimit() ) {\n'
-              + '  if ( sub != null ) sub.detach();\n'
-              + '} else {'
-              + '  setCount(getCount() + 1);\n'
-              + '  getDelegate().put(obj, sub);\n'
-              + '}\n'
+      javaCode:
```
Comment (kgrgreer) [inline src/foam/dao/SinkJava.js]: FYI: Best not to mix formatting and code changes in same PR. Then for the formatting changes are easier to review.

File: src/foam/lib/json/AnyParser.java [hunk 4]
```diff
@@ -18,6 +18,6 @@ public AnyParser() {
         new BooleanParser(),
         new LongParser(),
         new DoubleParser(),
-        new FObjectParser()));
+        new ExprParser()));
```
Comment (kgrgreer) [inline src/foam/lib/json/AnyParser.java]: What is this change for? What if I want to parse an FObject with the AnyParser?

File: src/foam/lib/json/FObjectArrayParser.java [hunk 5]
```diff
@@ -15,7 +15,7 @@ public FObjectArrayParser() {
                    new Literal("["),
                    new Whitespace(),
                    new Repeat(
-                              new FObjectParser(),
+                              new ExprParser(),
```
Comment (kgrgreer) [inline src/foam/lib/json/FObjectArrayParser.java]: Why?

File: src/foam/lib/json/JSONParser.java [hunk 6]
```diff
@@ -12,7 +12,7 @@
 public class JSONParser
   extends foam.core.ContextAwareSupport
 {
-  private Parser   parser   = new FObjectParser();
+  private Parser   parser   = new ExprParser();
```
Comment (kgrgreer) [inline src/foam/lib/json/JSONParser.java]: Use protected instead of private. Makes easier to test and debug. Why the change?

File: src/foam/lib/json/Outputter.java [hunk 7]
```diff
@@ -90,7 +90,9 @@ public void output(StringBuilder out, Object value) {
       outputFObject(out, (FObject)value);
     } else if ( value instanceof Number ) {
       outputNumber(out, (Number)value);
-    } else if ( value.getClass().isArray() ) {
+    } else if ( ( value != null ) &&
```
Comment (kgrgreer) [inline src/foam/lib/json/Outputter.java]: Move expression to an isArray() method.

File: src/foam/dao/SinkJava.js (hunk 2 above)
Comment (TanayParikh) [reply src/foam/dao/SinkJava.js]: Reverted multiline formatting for now

File: src/foam/dao/SinkJava.js (hunk 3 above)
Comment (TanayParikh) [reply src/foam/dao/SinkJava.js]: Thanks, noted. I've reverted the commit for now. I'll create a separate PR for it.

File: src/foam/lib/json/AnyParser.java (hunk 4 above)
Comment (TanayParikh) [reply src/foam/lib/json/AnyParser.java]: The `ExprParser` consists of an `Alt` with the two options being an `FObjectParser` and a custom parser for properties (`{ class: "__Property__", forClass_: "xyz" }`). This is needed to ensure properties are also parsed and handled correctly (the `__Property__` class isn't available from `foam.lookup()`, so the `FObjectParser` crashes if left as is).

File: src/foam/lib/json/JSONParser.java (hunk 6 above)
Comment (TanayParikh) [reply src/foam/lib/json/JSONParser.java]: Updated. 

The `ExprParser` consists of an `Alt` with the two options being an `FObjectParser` and a custom parser for properties (`{ class: "__Property__". forClass_: "xyz" }`). This is needed to ensure properties are also parsed and handled correctly (the `__Property__` class isn't available from `foam.lookup()`, so the `FObjectParser` crashes if left as is).

File: src/foam/lib/json/FObjectArrayParser.java (hunk 5 above)
Comment (TanayParikh) [reply src/foam/lib/json/FObjectArrayParser.java]: The `ExprParser` consists of an `Alt` with the two options being an `FObjectParser` and a custom parser for properties (`{ class: "__Property__", forClass_: "xyz" }`). This is needed to ensure properties are also parsed and handled correctly (the `__Property__` class isn't available from `foam.lookup()`, so the `FObjectParser` crashes if left as is).

File: src/foam/dao/AbstractDAO.java (hunk 1 above)
Comment (TanayParikh) [reply src/foam/dao/AbstractDAO.java]: Updated, **possible breaking change**. Fixed one internal issues, could impact other users.

https://github.com/foam-framework/foam2/pull/558/commits/d6d001cd5c50b9cc6157f27587de787f59862059

File: src/foam/lib/json/Outputter.java (hunk 7 above)
Comment (TanayParikh) [reply src/foam/lib/json/Outputter.java]: Updated.

File: src/foam/lib/json/AnyParser.java (hunk 4 above)
Comment (kgrgreer) [reply src/foam/lib/json/AnyParser.java]: Then, add a comment to explain that here or else add each of those Alt choices here directly.

File: src/foam/lib/json/FObjectArrayParser.java (hunk 5 above)
Comment (kgrgreer) [reply src/foam/lib/json/FObjectArrayParser.java]: Add comment.

File: src/foam/lib/json/AnyParser.java (hunk 4 above)
Comment (TanayParikh) [reply src/foam/lib/json/AnyParser.java]: So remove the existing `ExprParser`, make a new parser for `__Property__` and then replace L:21 with `new Alt(new __PropertyParser__(), new FObjectParser());`?

Or add the `Alt` choices in the comment?

Comment (kgrgreer) [reply src/foam/lib/json/AnyParser.java]: Yes, except L:21 is already an AltParser, so you don't need another one. Just call it PropertyParser.

File: src/foam/dao/AbstractDAO.java [hunk 8]
```diff
@@ -58,6 +58,8 @@ protected Sink decorateSink_(Sink sink, long skip, long limit, Comparator order,
       sink = new PredicatedSink(predicate, sink);
     }
 
+    ((AbstractFObject) sink).setX(x);
```
Comment (adamvy) [inline src/foam/dao/AbstractDAO.java]: What do we need this for?

Can we change the previous "new FooSink()" lines to read "x.create(FooSink.class)" instead?

File: src/foam/dao/SinkJava.js [hunk 9]
```diff
@@ -166,7 +166,7 @@ foam.CLASS({
     {
       name: 'eof',
       javaCode: 'if ( getArray() == null ) setArray(new java.util.ArrayList());\n'
-                + 'java.util.Collections.sort(getArray());\n'
+                + 'java.util.Collections.sort(getArray(), comparator_);\n'
```
Comment (adamvy) [inline src/foam/dao/SinkJava.js]: Use getComparator() instead.  It's preferable to access properties through the getFoo() interface, even if they are local properties so we don't skip any additional getter logic that FOAM properties may provide.

File: src/foam/dao/AbstractDAO.java (hunk 8 above)
Comment (TanayParikh) [reply src/foam/dao/AbstractDAO.java]: This is mainly to stay consistent with the other methods within the abstract class which accept a context. Ex. `pipe`, `removeAll_`.

File: src/foam/dao/SinkJava.js (hunk 9 above)
Comment (TanayParikh) [reply src/foam/dao/SinkJava.js]: Updated, thanks

File: src/foam/dao/AbstractDAO.java (hunk 8 above)
Comment (adamvy) [reply src/foam/dao/AbstractDAO.java]: I see.

I feel like we should be using x.create() in all those places rather than new Foo(), but I will get caught up on that code to see if we did it this way for any particular reason.

In this case in particular, setX() will only set the context on the latest decorated sink, but if we did a skipped, ordered, predicated sink, the skip and ordered Sinks will not have their context set.  I'm okay if we leave it as new Foo() but add the setX() calls after each new Foo() statement.

Comment (TanayParikh) [reply src/foam/dao/AbstractDAO.java]: I've updated accordingly for now. https://github.com/foam-framework/foam2/pull/558/commits/1af82c0f78fca2ed2c01401be2c53e0169a2cc47

Comment (kgrgreer) [reply src/foam/dao/AbstractDAO.java]: I don't want to incur the extra overhead of calling x.create() for internal functionality where I may never need it. Especially now that the Context hasn't been heavily optimized yet. Also, I'm not sure if I want people to redefine underlying predicates like EQ for security reasons.

Comment (adamvy) [reply src/foam/dao/AbstractDAO.java]: Went back if forth in my head for a while.  I get what you're saying.  Although in this context people couldn't redefine EQ() to have any ill effect since your PredicatedDAOs are already created, but they could redefine PredicatedSink.

In general I think things like DAOs which are both Context Aware and Context Oriented, should use the Context-Aware context for creating internal things, and not their context oriented context, which my suggestion would not be doing.

In this particular case, I think we should remove setX(), and also update OrderedSink to use "new foam.dao.Subscription()" rather than getX().create(foam.dao.Subscription.class); and everything should work for now.

File: src/foam/lib/json/FObjectArrayParser.java (hunk 5 above)
Comment (TanayParikh) [reply src/foam/lib/json/FObjectArrayParser.java]: Added

File: src/foam/dao/AbstractDAO.java (hunk 8 above)
Comment (TanayParikh) [reply src/foam/dao/AbstractDAO.java]: Updated.
